File: backend/scripts/send_warranty_alerts.py
"""Warranty expiry email alert — designed for cPanel Cron Jobs.

Standalone by design: opens its own DB connection, reads config from
backend/.env, and exits when done. It is NOT part of the web app process
(Passenger kills idle workers, so in-process schedulers like APScheduler
never fire reliably on shared hosting).

cPanel cron example (daily 08:00):
    0 8 * * * /home/USER/virtualenv/pc-vault-api/3.10/bin/python /home/USER/pc-vault-api/scripts/send_warranty_alerts.py >> /home/USER/warranty_cron.log 2>&1
"""
import smtplib
import sys
from datetime import date, timedelta
from email.message import EmailMessage
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from sqlalchemy import create_engine, select  # noqa: E402
from sqlalchemy.orm import Session, selectinload  # noqa: E402

from app.config import settings  # noqa: E402
from app.models import Part  # noqa: E402


def collect_expiring(session: Session, days: int) -> list[Part]:
    today = date.today()
    cutoff = today + timedelta(days=days)
    return list(
        session.execute(
            select(Part)
            .options(selectinload(Part.pc))
            .where(Part.warranty_expiry.is_not(None))
            .where(Part.warranty_expiry >= today)
            .where(Part.warranty_expiry <= cutoff)
            .order_by(Part.warranty_expiry.asc())
        ).scalars()
    )


def build_email(parts: list[Part], days: int) -> EmailMessage:
    today = date.today()
    lines = [
        f"PC Vault — {len(parts)} part(s) with warranty expiring within {days} days:",
        "",
    ]
    for p in parts:
        left = (p.warranty_expiry - today).days
        location = p.pc.name if p.pc else "inventory"
        lines.append(
            f"  • {p.brand} {p.model} ({p.type.value.upper()}) — "
            f"expires {p.warranty_expiry.isoformat()} ({left}d left) — in {location}"
        )
    lines += ["", f"Review: {settings.frontend_url.rstrip('/')}/alerts"]

    msg = EmailMessage()
    msg["Subject"] = f"[PC Vault] {len(parts)} warranty alert(s)"
    msg["From"] = settings.smtp_from
    msg["To"] = settings.alert_email_to
    msg.set_content("\n".join(lines))
    return msg


def send(msg: EmailMessage) -> None:
    if settings.smtp_port == 465:
        server = smtplib.SMTP_SSL(settings.smtp_host, settings.smtp_port, timeout=30)
    else:
        server = smtplib.SMTP(settings.smtp_host, settings.smtp_port, timeout=30)
        server.starttls()
    try:
        if settings.smtp_user:
            server.login(settings.smtp_user, settings.smtp_password)
        server.send_message(msg)
    finally:
        server.quit()


def main() -> None:
    if not settings.alert_email_to:
        print("ALERT_EMAIL_TO is not set — nothing to do.")
        return
    engine = create_engine(settings.database_url, pool_pre_ping=True)
    with Session(engine) as session:
        parts = collect_expiring(session, settings.warranty_alert_days)
    if not parts:
        print(f"{date.today().isoformat()}: no warranties expiring within "
              f"{settings.warranty_alert_days} days.")
        return
    msg = build_email(parts, settings.warranty_alert_days)
    if "--dry-run" in sys.argv:
        print(msg)
        return
    send(msg)
    print(f"{date.today().isoformat()}: sent alert for {len(parts)} part(s) "
          f"to {settings.alert_email_to}.")


if __name__ == "__main__":
    main()

File: backend/scripts/sendWarrantyAlerts.js
import 'dotenv/config';
import nodemailer from 'nodemailer';
import { connectDB } from '../config/db.js';
import { Part, PC } from '../config/models.js';

async function main() {
  await connectDB();

  const dryRun = process.argv.includes('--dry-run');
  const days = parseInt(process.env.WARRANTY_ALERT_DAYS || '30', 10);
  const today = new Date();
  const cutoff = new Date(today.getTime() + days * 86400000);

  let parts = await Part.find({}).lean();
  parts = parts.filter(p => p.warranty_expiry && new Date(p.warranty_expiry) >= today && new Date(p.warranty_expiry) <= cutoff);
  parts.sort((a, b) => new Date(a.warranty_expiry) - new Date(b.warranty_expiry));

  if (parts.length === 0) {
    console.log('No warranties expiring within', days, 'days.');
    process.exit(0);
  }

  const enriched = [];
  for (const part of parts) {
    let location = 'inventory';
    if (part.pc_id) {
      const pc = await PC.findById(part.pc_id).lean();
      location = pc ? `PC: ${pc.name}` : 'unknown PC';
    } else if (part.employee_id) {
      location = 'assigned to employee';
    }
    enriched.push({ ...part, location });
  }

  const alertEmail = process.env.ALERT_EMAIL_TO;
  if (!alertEmail) {
    console.log('ALERT_EMAIL_TO not set — printing results:');
    for (const p of enriched) {
      console.log(`  ${p.brand} ${p.model} (SN: ${p.serial_number || '—'}) — expires ${new Date(p.warranty_expiry).toISOString().split('T')[0]} (${p.location})`);
    }
    process.exit(0);
  }

  const transporter = nodemailer.createTransport({
    host: process.env.SMTP_HOST || 'localhost',
    port: parseInt(process.env.SMTP_PORT || '465', 10),
    secure: parseInt(process.env.SMTP_PORT || '465', 10) === 465,
    auth: process.env.SMTP_USER
      ? { user: process.env.SMTP_USER, pass: process.env.SMTP_PASSWORD }
      : undefined,
  });

  const lines = enriched.map(
    (p) => `  • ${p.brand} ${p.model} (SN: ${p.serial_number || '—'}) — warranty ends ${new Date(p.warranty_expiry).toISOString().split('T')[0]} — ${p.location}`,
  );

  const mail = {
    from: process.env.SMTP_FROM || 'pc-vault@localhost',
    to: alertEmail,
    subject: `PC Vault — ${parts.length} warranty(ies) expiring within ${days} days`,
    text: [
      `The following ${parts.length} item(s) have warranties expiring in the next ${days} days:\n`,
      ...lines,
      '\n— PC Vault',
    ].join('\n'),
  };

  if (dryRun) {
    console.log('=== DRY RUN — no email sent ===');
    console.log('To:', mail.to);
    console.log('Subject:', mail.subject);
    console.log(mail.text);
  } else {
    await transporter.sendMail(mail);
    console.log('Alert email sent to', alertEmail);
  }

  process.exit(0);
}

main().catch((err) => {
  console.error('Warranty alerts failed:', err);
  process.exit(1);
});
